src/mlpack/methods/hoeffding_trees/hoeffding_tree.hpp

```diff
@@ -194,44 +265,136 @@ class HoeffdingTree
-             const bool batchTraining = true,
-             const bool resetTree = false,
-             const size_t numClasses = 0);
+             const size_t numClasses,
+             const bool batchTraining,
+             const double successProbability,
+             const size_t maxSamples,
+             const size_t checkInterval,
+             const size_t minSamples);
 
   /**
    * Train on a set of points, either in streaming mode or in batch mode, with
-   * the given labels and the given `DatasetInfo`.  This will reset the tree.
-   * This only needs to be called when the `DatasetInfo` has changed---if you
-   * are training incrementally but have already passed the DatasetInfo once,
-   * use the overload of `Train()` that does not take a `DatasetInfo` and make
-   * sure `resetTree` is set to `false`.
+   * the given labels.  If `resetTree` is set to `true`, then reset the state of
+   * the tree to an empty tree before training.
+   *
+   * Note that the tree will be automatically reset if the dimensionality of
+   * `data` does not match the dimensionality that the tree was currently
+   * trained with.  The tree will also be reset if `numClasses` is passed and
+   * differs from the existing setting.
    *
    * @param data Data points to train on.
-   * @param info DatasetInfo object with information about each dimension.
+   * @param datasetInfo Information on the dataset (types of each feature).
    * @param labels Labels of data points.
+   * @param numClasses The number of classes in `labels`.  Passing this will
+   *      reset the tree.  If not given and `resetTree` is `true`, then the
+   *      number of classes will be computed from `labels`.
    * @param batchTraining If true, perform training in batch.
-   * @param numClasses Number of classes in `labels`.  If not specified, it is
-   *      computed from `labels`.
+   * @param successProbability Probability of success required in Hoeffding
+   *      bounds before a split can happen.
+   * @param maxSamples Maximum number of samples before a split is forced (0
+   *      never forces a split); ignored in batch training mode.
+   * @param checkInterval Number of samples required before each split; ignored
+   *      in batch training mode.
+   * @param minSamples If the node has seen this many points or fewer, no split
+   *      will be allowed.
    */
+  // Many overloads needed here until we have std::optional.
+  template<typename MatType>
+  void Train(const MatType& data,
+             const data::DatasetInfo& info,
+             const arma::Row<size_t>& labels,
+             const size_t numClasses = 0,
+             const bool batchTraining = true);
+
+  template<typename MatType>
+  void Train(const MatType& data,
+             const data::DatasetInfo& info,
+             const arma::Row<size_t>& labels,
+             const size_t numClasses,
+             const bool batchTraining,
+             const double successProbability);
```

Comment:
Can we make the `sucessProbability` as ElemType?

Reply:
I didn't do this, mostly so that when you switch to using a different `ElemType`, as few things change in the API as possible.  (That will also make the documentation job in #3584 easier.)  For something like the success probability, using a `double` or a `float` or whatever other element type really won't make any serious difference, so I think it might be easier to leave as-is.